### PR TITLE
Fix transaction rollback if `WriteBuffer::finalize` fails in plain_rewritable disk during directory creation

### PR DESCRIPTION
**Pull request by jkartseva**
Fixes https://github.com/ClickHouse/clickhouse-private/issues/18866

Do not store a directory path in the in-memory map if `WriteBuffer::finalize` fails
```
2024.11.01 20:34:14.237649 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Trace> MetadataStorageFromPlainObjectStorageCreateDirectoryOperation: Creating metadata for directory 'store/59d/59dd1264-ed0a-4276-9e9e-85a424e09f7a/tmp_insert_3_5_5_0/
' with remote path='03008_test_s3_mt/__meta/jwujyiezwuprtdvwujlndrgieeblyyug/prefix.path'
2024.11.01 20:34:14.239346 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Warning> WriteBufferFromS3: Nothing to abort. Details: bucket test, key 03008_test_s3_mt/__meta/jwujyiezwuprtdvwujlndrgieeblyyug/prefix.path, total size 66, count 0, hid
den_size 0, offset 0, with pool: true, prefinalized true, finalized false
2024.11.01 20:34:14.240731 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Information> WriteBufferFromS3: WriteBufferFromS3 was canceled.The file might not be written to S3. Details: bucket test, key 03008_test_s3_mt/__meta/jwujyiezwuprtdvwujl
ndrgieeblyyug/prefix.path, total size 66, count 0, hidden_size 0, offset 0, with pool: true, prefinalized true, finalized false.
```
```
2024.11.01 20:34:14.608171 [ 14798 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Error> executeQuery: Code: 439. DB::Exception: Cannot schedule a task: fault injected (threads=2, jobs=0): While committing metadata operation #0. (CANNOT_SCHEDULE_TASK) (version 24.11.1.11636 (official build)) (from [::1]:46576) (comment: 03008_s3_plain_rewritable.sh) (in query: insert into test_s3_mt (*) select number, number, number from numbers_mt(10000);), Stack trace (when copying this message, always include the lines below):

0. /build/contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x0000000016ec78f2
1. /build/src/Common/Exception.cpp:109: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c5679d9
2. /build/src/Common/Exception.h:110: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000722e20c
3. /build/src/Common/Exception.h:128: DB::Exception::Exception<String const&, unsigned long, unsigned long&>(int, FormatStringHelperImpl<std::type_identity<String const&>::type, std::type_identity<unsigned long>::type, std::type_identity<unsigned long&>::type>, String const&, unsigned long&&, unsigned long&) @ 0x000000000c62128b
4. /build/src/Common/ThreadPool.cpp:259: void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda'(String const&)::operator()(String const&) const @ 0x000000000c622ab3
5. /build/src/Common/ThreadPool.cpp:307: void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool) @ 0x000000000c61e5ac
6. /build/src/Common/ThreadPool.cpp:478: ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleOrThrowOnError(std::function<void ()>, Priority) @ 0x000000000c61e40e
7. /build/src/Common/threadPoolCallbackRunner.h:59: std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunnerUnsafe<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, String const&)::'lambda'(std::function<void ()>&&, Priority)::operator()(std::function<void ()>&&, Priority) @ 0x00000000108e2e47
8. /build/contrib/llvm-project/libcxx/include/__functional/invoke.h:394: ? @ 0x00000000108e2c6f
9. /build/contrib/llvm-project/libcxx/include/__functional/function.h:848: ? @ 0x000000001096456b
10. /build/src/IO/WriteBufferFromS3.cpp:772: DB::WriteBufferFromS3::makeSinglepartUpload(DB::WriteBufferFromS3::PartData&&) @ 0x0000000010952e89
11. /build/src/IO/WriteBufferFromS3.cpp:193: DB::WriteBufferFromS3::preFinalize() @ 0x0000000010952835
12. /build/src/IO/WriteBufferFromS3.cpp:213: DB::WriteBufferFromS3::finalizeImpl() @ 0x000000001095354a
13. /build/src/IO/WriteBuffer.h:131: DB::WriteBuffer::finalize() @ 0x0000000007239b90
14. /build/src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp:88: DB::MetadataStorageFromPlainObjectStorageCreateDirectoryOperation::execute(std::unique_lock<DB::SharedMutex>&) @ 0x00000000112d6a83
15. /build/src/Disks/ObjectStorages/MetadataOperationsHolder.cpp:65: DB::MetadataOperationsHolder::commitImpl(std::variant<std::monostate, DB::MetaInKeeperCommitOptions<zkutil::ZooKeeper>, DB::MetaInKeeperCommitOptions<DB::ZooKeeperWithFaultInjection>> const&, DB::SharedMutex&) @ 0x00000000112701bc
16. /build/src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:1020: DB::DiskObjectStorageTransaction::commit(std::variant<std::monostate, DB::MetaInKeeperCommitOptions<zkutil::ZooKeeper>, DB::MetaInKeeperCommitOptions<DB::ZooKeeperWithFaultInjection>> const&) @ 0x000000001121d7b2
17. /build/src/Disks/ObjectStorages/DiskObjectStorage.cpp:394: DB::DiskObjectStorage::createDirectories(String const&) @ 0x000000001120a466
18. /build/src/Storages/MergeTree/DataPartStorageOnDiskBase.cpp:1103: DB::DataPartStorageOnDiskBase::createDirectories() @ 0x0000000012eff925
19. /build/src/Storages/MergeTree/MergeTreeDataWriter.cpp:653: DB::MergeTreeDataWriter::writeTempPartImpl(DB::BlockWithPartition&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::Context const>, long, bool) @ 0x00000000130d65b8
```
...because it leads to `DB::Exception: The specified key does not exist. This error happened for S3 disk.` error:
```
2024.11.01 20:34:14.405961 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Trace> MetadataStorageFromPlainObjectStorageMoveDirectoryOperation: Moving directory 'store/59d/59dd1264-ed0a-4276-9e9e-85a424e09f7a/tmp_insert_3_5_5_0/' to 'store/59d/5
9dd1264-ed0a-4276-9e9e-85a424e09f7a/delete_tmp_tmp_insert_3_5_5_0/'
2024.11.01 20:34:14.407800 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Information> AWSClient: Response status: 404, Not Found
2024.11.01 20:34:14.408483 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Information> AWSClient: AWSErrorMarshaller: Encountered AWSError 'NoSuchKey': The specified key does not exist.
2024.11.01 20:34:14.409018 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Information> AWSClient: AWSXmlClient: HTTP response code: 404
Resolved remote host IP address: 127.0.0.1:11111
Request ID: 1803EF2C127189A7
Exception name: NoSuchKey
Error message: The specified key does not exist.
11 response headers:
accept-ranges : bytes
--
content-type : application/xml
date : Fri, 01 Nov 2024 19:34:14 GMT
server : MinIO
strict-transport-security : max-age=31536000; includeSubDomains
vary : Accept-Encoding
x-amz-id-2 : 7dc7eb22d3288ec80374614e9088e31d3668a6922ead55932dd2a8e56373820f
x-amz-request-id : 1803EF2C127189A7
x-content-type-options : nosniff
x-xss-protection : 1; mode=block
2024.11.01 20:34:14.411105 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Information> AWSClient: If the signature check failed. This could be because of a time skew. Attempting to adjust the signer.
2024.11.01 20:34:14.411799 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Debug> ReadBufferFromS3: Caught exception while reading S3 object. Bucket: test, Key: 03008_test_s3_mt/__meta/jwujyiezwuprtdvwujlndrgieeblyyug/prefix.path, Version: Latest, Offset: 0, Attempt: 1/4, Message: Code: 499. DB::Exception: The specified key does not exist. This error happened for S3 disk. (S3_ERROR) (version 24.11.1.11636 (official build))
2024.11.01 20:34:14.420090 [ 4887 ] {} <Trace> SystemLog (system.blob_storage_log): Flushing system log, 12 entries to flush up to offset 15
2024.11.01 20:34:14.421763 [ 4887 ] {} <Debug> SystemLog (system.blob_storage_log): Will ignore blobs with prefix store/531/531eb1b8-b03a-4536-9cc5-6eb22d42f41e
2024.11.01 20:34:14.437289 [ 4887 ] {} <Trace> Planner: Query to stage Complete only analyze
2024.11.01 20:34:14.438034 [ 4887 ] {} <Trace> Planner: Query from stage FetchColumns to stage Complete only analyze
2024.11.01 20:34:14.439139 [ 5753 ] {ae7285ba-a64c-4da8-b10c-7ee11dde90b7} <Error> void DB::MetadataOperationsHolder::commitImpl(const TransactionCommitOptionsVariant &, SharedMutex &): Code: 499. DB::Exception: The specified key does not exist.
 This error happened for S3 disk.: while reading key: 03008_test_s3_mt/__meta/jwujyiezwuprtdvwujlndrgieeblyyug/prefix.path, from bucket: test. (S3_ERROR), Stack trace (when copying this message, always include the lines below):

0. /build/contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x0000000016ec78f2
1. /build/src/Common/Exception.cpp:109: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c5679d9
2. /build/src/Common/Exception.h:109: DB::S3Exception::S3Exception(String const&, Aws::S3::S3Errors) @ 0x00000000108e4132
3. /build/src/IO/ReadBufferFromS3.cpp:439: DB::ReadBufferFromS3::sendRequest(unsigned long, unsigned long, std::optional<unsigned long>) const @ 0x000000001094cc93
4. /build/src/IO/ReadBufferFromS3.cpp:394: DB::ReadBufferFromS3::initialize(unsigned long) @ 0x000000001094b396
5. /build/src/IO/ReadBufferFromS3.cpp:128: DB::ReadBufferFromS3::nextImpl() @ 0x000000001094af7b
6. /build/src/IO/ReadBuffer.h:70: DB::ReadBuffer::next() @ 0x0000000007a5243b
7. /build/src/IO/ReadBuffer.h:106: DB::readStringUntilEOF(String&, DB::ReadBuffer&) @ 0x000000000c5c851d
8. /build/src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp:161: DB::MetadataStorageFromPlainObjectStorageMoveDirectoryOperation::createWriteBuf(std::__fs::filesystem::path const&, std::__fs::filesystem::path const&, b
ool) @ 0x00000000112d84ff
```
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix NoSuchKey error during transaction rollback when creating a directory fails for the palin_rewritable disk.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit de18a667df48c47ebffce5d369201c25c168af9b with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71439/de18a667df48c47ebffce5d369201c25c168af9b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/71402/5c69cf3205c77bbb2dcf8a5d52539679507c86fe/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp [hunk 1]
```diff
@@ -97,20 +97,25 @@ void MetadataStorageFromPlainObjectStorageCreateDirectoryOperation::undo(std::un
 {
     auto metadata_object_key = createMetadataObjectKey(object_key_prefix, metadata_key_prefix);
 
-    if (write_finalized)
+    if (write_finalized || write_created)
```
Comment (tuanpach) [inline src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp:100]: From what I understand, we need to remove the `parent_path` from the `path_map` when `write_created` too.

Can we add a new method `removePath` in `InMemoryDirectoryPathMap` to remove the entry?

The following is my suggestion:

```C++
struct InMemoryDirectoryPathMap
{
    /**
     * @return the number of removed entries
     */
    size_t removePath(const std::filesystem::path & path)
    {
        std::lock_guard lock(mutex);
        return map.erase(path);
    }
}

void MetadataStorageFromPlainObjectStorageCreateDirectoryOperation::undo(std::unique_lock<SharedMutex> &)
{    
    if (!write_finalized && !write_created)
        return;

    size_t erase_count = path_map.removePath(path.parent_path());
    if (erase_count)
    {
       // Should we use `auto & metric` here?
        auto metric = object_storage->getMetadataStorageMetrics().directory_map_size;
        CurrentMetrics::sub(metric, erase_count);
    }

    StoredObject stored_object{metadata_object_key.serialize(), path / PREFIX_PATH_FILE_NAME};
    if (write_finalized)
        object_storage->removeObject(stored_object);
    else // write_created
        object_storage->removeObjectIfExists(stored_object);
}
```

Comment (jkartseva) [reply src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp:100]: Thanks, I'll address it.

**Comment by jkartseva** [comment]
Having a symmetric `addPath` method would be nice, but it is a refactoring and is beyond the scope of the fix. I'll create a follow-up.

**Comment by tuanpach** [comment]
@jkartseva The changes LGTM.

Should we add some tests reproducing the issue in the description? And verify that the new logic works well.

**Review by tuanpach — APPROVED** [review]

**Pull request merged**